### PR TITLE
Add multivariate normal distribution

### PR DESCRIPTION
**Pull request by YoshikawaMasashi**
split from #3622 .

I have not implemented `sample` and `mean` test yet, because `mean` and `sample` method is invalid in `scipy.stats multivariate_normal`. Therefore, we should implement these tests in a different way.

### CODE REVIEW

File: chainer/distributions/multivariate_normal.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    @property
+    def batch_shape(self):
+        return self.loc.shape[:-1]
+
+    @property
+    def entropy(self):
+        st = rollaxis.rollaxis(self.scale_tril, -2, 0)
+        st = rollaxis.rollaxis(st, -1, 1)
+        diag = st[list(range(self.d)), list(range(self.d))]
```
Comment (toslunar) [inline chainer/distributions/multivariate_normal.py]: I'll implement `F.diagonal`.

File: chainer/distributions/multivariate_normal.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    @property
+    def _is_gpu(self):
+        return isinstance(self.loc.data, cuda.ndarray)
+
+    def log_prob(self, x):
+        st = rollaxis.rollaxis(self.scale_tril, -2, 0)
+        st = rollaxis.rollaxis(st, -1, 1)
+        diag = st[list(range(self.d)), list(range(self.d))]
+        logdet = sum.sum(exponential.log(basic_math.absolute(diag)), axis=0)
```
Comment (toslunar) [inline chainer/distributions/multivariate_normal.py]: Why don't you add `_logdet` method, which would be called by `entropy` and `log_prob`?

File: chainer/distributions/multivariate_normal.py (hunk 1 above)
Comment (YoshikawaMasashi) [reply chainer/distributions/multivariate_normal.py]: `F.Cholesky` may also become necessary for full covariance (v.s. `scale_tril`). I may implement this.

**Comment by toslunar** [comment]
It seems we have no specialization for computing inverse of a *lower triangular* matrix in GPU (`chainer.functions.math.inv._inv_gpu`, `cupy.linalg.inv`).

**Comment by YoshikawaMasashi** [comment]
Oh, OK.....
I'll try to implement `F.solve_ solve_triangular` and `cupyx.scipy.linalg.solve_triangular` as like `scipy.linalg.solve_triangular`.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve_triangular.html

**Comment by YoshikawaMasashi** [comment]
I implemented `cupyx.scipy.linalg.solve_triangular`. But `scipy.linalg.solve_triangular` and `cupyx.scipy.linalg.solve_triangular` can deal with 2d-matrix and cannot deal with batched processing...
Is there any good way?...

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

File: tests/chainer_tests/distributions_tests/test_multivariate_normal.py [hunk 3]
```diff
@@ -40,4 +45,72 @@ def sample_for_test(self):
...
+
+    def test_double_backward_cpu(self):
+        self.check_double_backward(self.x, self.gy, self.ggy)
+
+    @attr.gpu
+    def test_double_backward_gpu(self):
+        self.check_double_backward(
+            cuda.to_gpu(self.x), cuda.to_gpu(self.gy), cuda.to_gpu(self.ggy))
+
+
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_multivariate_normal.py]: Could you add tests `without_requires('scipy')`, as in `tests/chainer_tests/functions_tests/math_tests/test_lgamma.py`?

File: tests/chainer_tests/distributions_tests/test_multivariate_normal.py [hunk 4]
```diff
@@ -40,4 +45,72 @@ def sample_for_test(self):
...
+        self.ggy = numpy.random.normal(
+            size=(self.d, self.d)).astype(self.dtype)
+        self.backward_options = {'atol': 1e-2, 'rtol': 1e-2}
+        self.double_backward_options = {'atol': 1e-2, 'rtol': 1e-2}
+
+    def check_forward(self, x_data):
+        xp = cuda.get_array_module(x_data)
+        y = distributions.multivariate_normal._triangular_inv(
+            x_data, lower=self.lower)
+        if xp == cuda.cupy:
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_multivariate_normal.py]: It's OK here to trust `cupy.linalg.inv`.  Could you just do `xp.linalg.inv`?

File: tests/chainer_tests/distributions_tests/test_multivariate_normal.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+class TestMultivariateNormal(testing.distribution_unittest):
+
+    scipy_onebyone = True
+
+    def setUp_configure(self):
+        from scipy import stats
+        self.dist = distributions.MultivariateNormal
+        self.scipy_dist = stats.multivariate_normal
+        self.scipy_onebyone = True
+        self.event_shape = (3,)
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_multivariate_normal.py]: Could you make this a testing parameter?
```
    'event_shape': [(3,)],
```
```
    d, = self.event_shape
```

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

File: chainer/distributions/multivariate_normal.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+from chainer.functions.array import swapaxes
+from chainer.functions.array import where
+from chainer.functions.math import basic_math
+from chainer.functions.math import exponential
+from chainer.functions.math import matmul
+from chainer.functions.math import sum as sum_mod
+from chainer import utils
+from chainer.utils import type_check
+import math
+import numpy
```
Comment (toslunar) [inline chainer/distributions/multivariate_normal.py]: Could you fix the import order?



> 1. Standard library imports.
> 2. Related third party imports.
> 3. Local application/library specific imports.

https://www.python.org/dev/peps/pep-0008/

File: chainer/distributions/multivariate_normal.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+        invx, = self.get_retained_outputs()
+        mask = xp.tril(xp.ones((len(invx), len(invx)), dtype=bool))
+        if not self._lower:
+            mask = mask.T
+        # Gradient is - x^-T (dx) x^-T
+        invxT = chainer.functions.transpose(invx)
+        gx = chainer.functions.matmul(
+            chainer.functions.matmul(- invxT, gy), invxT)
+        gx = where.where(mask, gx, xp.zeros_like(gx.array))
+        print(gx)
```
Comment (toslunar) [inline chainer/distributions/multivariate_normal.py]: Remove the debug print.

File: chainer/distributions/multivariate_normal.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    Args:
+        loc(:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+        :class:`cupy.ndarray`): Parameter of distribution representing the \
+        location :math:`\\mu`.
+        scale_tril(:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
+        :class:`cupy.ndarray`): Parameter of distribution representing the \
+        scale :math:`L`.
+    """
+
+    def __init__(self, loc, scale_tril):
```
Comment (toslunar) [inline chainer/distributions/multivariate_normal.py]: Could you make `scale_tril` a keyword-only argument, as we might support `cov` in future?

File: chainer/distributions/multivariate_normal.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        type_check.expect(a_type.shape[0] == a_type.shape[1])
+
+    def forward_cpu(self, inputs):
+        self.retain_outputs((0,))
+        if not available_cpu:
+            raise ImportError("SciPy is not available. Forward computation"
+                              " of triangular_inv in CPU can not be done." +
+                              str(_import_error))
+        x, = inputs
+        invx = scipy.linalg.solve_triangular(x, numpy.eye(len(x)),
```
Comment (toslunar) [inline chainer/distributions/multivariate_normal.py]: How about giving `dtype` to `numpy.eye`?  (Then the casting by `force_array` could be avoided.)

File: chainer/distributions/multivariate_normal.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+                                             lower=self._lower)
+        return utils.force_array(invx, dtype=x[0].dtype),
+
+    def forward_gpu(self, inputs):
+        self.retain_outputs((0,))
+        x, = inputs
+        invx = cuda.cupyx.scipy.linalg.solve_triangular(
+            x, cuda.cupy.eye(len(x)), lower=self._lower)
```
Comment (toslunar) [inline chainer/distributions/multivariate_normal.py]: Ditto. (See the comment in `forward_cpu`.)

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by YoshikawaMasashi** [comment]
jenkins, test this please.

**Comment by toslunar** [comment]
The Jenkins failure (`tests.chainer_tests.functions_tests.normalization_tests.test_l2_normalization.TestL2Normalization_param_50.test_backward_gpu`) is irrelevant to the PR.

**Pull request merged**